### PR TITLE
Refactor options

### PR DESCRIPTION
**Pull request by Fatme**

### CODE REVIEW

File: commands/device/list-applications.ts [hunk 1]
```diff
@@ -1,20 +1,21 @@
...
 
 	allowedParameters: ICommandParameter[] = [];
 
 	public execute(args: string[]): IFuture<void> {
 		return (() => {
-			this.$devicesServices.initialize({ deviceId: options.device, skipInferPlatform: true }).wait();
-			let output: string[] = [];
+
+			this.$devicesServices.initialize({ deviceId: this.$options.device, skipInferPlatform: true }).wait();
+			var output: string[] = [];
```
Comment (rosen-vladimirov) [inline commands/device/list-applications.ts]: use let :)

File: commands/device/list-devices.ts [hunk 2]
```diff
@@ -17,8 +17,8 @@ export class ListDevicesCommand implements ICommand {
 			let index = 1;
 			this.$devicesServices.initialize({platform: args[0], deviceId: null, skipInferPlatform: true}).wait();
 
-			let action: (device: Mobile.IDevice) => IFuture<void>;
-			if (options.json) {
+			var action: (device: Mobile.IDevice) => IFuture<void>;
```
Comment (rosen-vladimirov) [inline commands/device/list-devices.ts]: use let :)

File: errors.ts [hunk 3]
```diff
@@ -64,7 +64,7 @@ function resolveCallStack(error: Error): string {
 }
 
 export function installUncaughtExceptionListener(): void {
-	process.on("uncaughtException", (err: Error) => {
+	process.on("uncaughtException", (err: Error) => {
```
Comment (teobugslayer) [inline errors.ts]: Unneeded tab characters added

File: host-info.ts [hunk 4]
```diff
@@ -1,64 +1,80 @@
 ///<reference path="../.d.ts"/>
 "use strict";
-
 import Future = require("fibers/future");
 
-export function isWindows() {
-	return process.platform === "win32";
-}
+export class HostInfo implements IHostInfo {
+	private static WIN32 = "win32";
```
Comment (teobugslayer) [inline host-info.ts]: Why is this simply WIN32 but others are XXX_OS_NAME?

File: mobile/android/android-device.ts [hunk 5]
```diff
@@ -218,9 +219,10 @@ export class AndroidDevice implements Mobile.IAndroidDevice {
 
     private attachDebugger(packageName: string): void {
         let startDebuggerCommand = this.composeCommand("shell am broadcast -a \"%s-Debug\" --ez enable true", packageName);
-        let port = options.debugPort;
+        let port = this.$options.debugPort;
+		
         if (port > 0) {
-            startDebuggerCommand += " --ei debuggerPort " + options["debug-port"];
+            startDebuggerCommand += " --ei debuggerPort " + this.$options["debug-port"];
```
Comment (teobugslayer) [inline mobile/android/android-device.ts]: can we use ```this.$options.debugPort``` here?

File: mobile/android/android-device.ts [hunk 6]
```diff
@@ -252,7 +254,7 @@ export class AndroidDevice implements Mobile.IAndroidDevice {
         let installCommand = this.composeCommand("install -r \"%s\"", packageFile);
         this.$childProcess.exec(installCommand).wait();
 
-        let port = options["debug-port"];
+        let port = this.$options["debug-port"];
```
Comment (teobugslayer) [inline mobile/android/android-device.ts]: We can use the shorter ```this.$options.debugPort```

File: declarations.d.ts [hunk 7]
```diff
@@ -301,3 +302,60 @@ interface ISysInfoData {
...
+	json: boolean;
+	watch: boolean;
+	avd: string;
+	"profile-dir": string;
+	timeout: string;
+	device: string;
+	availableDevices: boolean;
+	appid: string;
+	geny: string;
+	"debug-brk": boolean;
```
Comment (teobugslayer) [inline declarations.d.ts]: can we define debugBrk as well?

File: declarations.d.ts [hunk 8]
```diff
@@ -301,3 +302,60 @@ interface ISysInfoData {
...
+	"debug-port": number;
+	debugPort: number;
+	"get-port": boolean;
```
Comment (teobugslayer) [inline declarations.d.ts]: can we define getPort as well?

File: mobile/android/android-device.ts [hunk 9]
```diff
@@ -275,13 +277,13 @@ export class AndroidDevice implements Mobile.IAndroidDevice {
 
     public debug(packageFile: string, packageName: string): IFuture<void> {
         return (() => {
-            if (options["get-port"]) {
+            if (this.$options["get-port"]) {
```
Comment (teobugslayer) [inline mobile/android/android-device.ts]: we could use getPort but it needs to be defined in the interface

File: mobile/android/android-device.ts [hunk 10]
```diff
@@ -275,13 +277,13 @@ export class AndroidDevice implements Mobile.IAndroidDevice {
...
                 this.printDebugPort(packageName);
-            } else if (options["start"]) {
+            } else if (this.$options["start"]) {
                 this.attachDebugger(packageName);
-            } else if (options["stop"]) {
+            } else if (this.$options["stop"]) {
                 this.detachDebugger(packageName);
-            } else if (options["debug-brk"]) {
+            } else if (this.$options["debug-brk"]) {
```
Comment (teobugslayer) [inline mobile/android/android-device.ts]: we could use debugBrk but it needs to be defined in the interface

File: options.ts [hunk 11]
```diff
@@ -1,67 +1,160 @@
...
-declare let exports: any;
+export enum OptionType {
+	String,
+	Boolean,
+	Number,
+	Array
+}
+
+export class OptionsBase {
+	private optionsWhiteList = ["ui", "recursive", "reporter", "require", "timeout", "_", "$0"]; // This options shouldn't be validated
```
Comment (teobugslayer) [inline options.ts]: This -> These

File: declarations.d.ts (hunk 7 above)
Comment (Fatme) [inline declarations.d.ts]: Yes, I've fixed this.

File: services/commands-service.ts [hunk 12]
```diff
@@ -79,11 +80,16 @@ export class CommandsService implements ICommandsService {
 
 	public tryExecuteCommand(commandName: string, commandArguments: string[]): IFuture<void> {
 		return (() => {
-			if(!this.areDynamicSubcommandsRegistered) {
-				this.$commandsServiceProvider.registerDynamicSubCommands();
-				this.areDynamicSubcommandsRegistered = true;
+			var  x = (commandName: string, commandArguments: string[]) => {
```
Comment (teobugslayer) [inline services/commands-service.ts]: The name "x" can be improved :) I guess this is left-over spiked code

**Comment by teobugslayer** [comment]
Awesome work! :+1:

**Comment by rosen-vladimirov** [comment]
:+1:

**Pull request merged**